### PR TITLE
Add tracer refinement criterion

### PR DESCRIPTION
**Pull request by gassmoeller**
The entry in changes.h says it all. This PR adds a mesh refinement plugin that refines cells based on the density of tracers in that cell. It can be used to balance the load between cells and particles more effectively, as long as one is fine with synchronizing particle density and cell size.

### CODE REVIEW

File: source/mesh_refinement/particle_density.cc [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+// explicit instantiations
+namespace aspect
+{
+  namespace MeshRefinement
+  {
+    ASPECT_REGISTER_MESH_REFINEMENT_CRITERION(ParticleDensity,
+                                              "particle density",
+                                              "A mesh refinement criterion that computes "
+                                              "refinement indicators from the density of particles.")
```
Comment (bangerth) [inline source/mesh_refinement/particle_density.cc]: I think this is too vague. What you want to say is `...computes refinement indicators that equal the areal (in 2d) or volume (in 3d) density of particles in this cell.`

**Comment by bangerth** [comment]
I don't know how this would be useful, but there's little harm to adding it. Please fix up the message and merge yourself.

**Comment by gassmoeller** [comment]
Think of a model with a passive compositional field, which has an interface you want to track. With compositional fields you would use a `composition` refinement indicator, but with tracers you would need something that looks up the properties of the particles. For now I postponed this more complicated plugin, but used the particle density instead (which you would want to be high close to the interface, if you know where the interface is). 

Plus: This plugin nicely balances the computational load between particles and mesh, if you have inhomogeneous particle distributions and are most interested in the region of high particle density anyway.

**Comment by bangerth** [comment]
Thanks for the explanation. Could you add essentially this to the description to give people an idea what the plugin may be useful for? Merge is pre-approved.

**Pull request merged**